### PR TITLE
Resurrect patches from unstable branch

### PR DESCRIPTION
**Pull request by CyberShadow**
I ran a [git filter-branch](https://gist.github.com/faa08346748a968c2c8561eb61c6afe0) on a part of the old unstable branch to make cherry-picking easier. I'm including mostly my own commits, but more commits from the unstable branch can thus be resurrected from here (let me know if you would like me to add any particular ones):
https://github.com/CyberShadow/bmo/commits/unstable-reformatted

The commits here have all been previously reviewed (see the pull request references added by GitHub in the commit messages) and theoretically don't need further review but practically need to be retested due to bitrot (or e.g. mistakes during merge conflict resolution).

Opening as draft until I can test this as part of a full test deployment.

### CODE REVIEW

**Comment by CyberShadow** [comment]
Basic functionality seems to be working.

**Comment by CyberShadow** [comment]
I'm seeing some test failures (some caused by patches in this PR, some not), will go through those first.

File: extensions/BMO/template/en/default/pages/bug-writing.html.tmpl [hunk 1]
```diff
@@ -7,5 +7,5 @@
   #%]
 
 [% PROCESS global/redirect.html.tmpl
-   url = "https://developer.mozilla.org/docs/Mozilla/QA/Bug_writing_guidelines"
+   url = "[% terms.BugWritingGuidelinesURL %]"
```
Comment (dylanwh) [inline extensions/BMO/template/en/default/pages/bug-writing.html.tmpl:10]:
```suggestion
   url = terms.BugWritingGuidelinesURL
```

Comment (CyberShadow) [reply extensions/BMO/template/en/default/pages/bug-writing.html.tmpl:10]: Yes, I saw you fixed this previously in 01732cefbaddddb93ad21adff590294799f5d2bc :) I will apply that here.

File: Bugzilla.pm [hunk 2]
```diff
@@ -124,6 +125,14 @@ sub template_inner {
 }
```
Comment (dylanwh) [inline Bugzilla.pm:126]:
```suggestion
our $in_extensions = 0;
```

File: Bugzilla.pm [hunk 3]
```diff
@@ -124,6 +125,14 @@ sub template_inner {
...
 
 sub extensions {
+  # Guard against extensions querying the extension list during initialization
+  # (through this method or has_extension).
+  # The extension list is not fully populated at that point,
+  # so the results would not be meaningful.
+  state $recursive = 0;
```
Comment (dylanwh) [inline Bugzilla.pm:132]:
```suggestion
```

File: Bugzilla.pm [hunk 4]
```diff
@@ -124,6 +125,14 @@ sub template_inner {
...
+  die "Recursive attempt to load/query extensions" if $recursive;
```
Comment (dylanwh) [inline Bugzilla.pm:133]:
```suggestion
  die "Recursive attempt to load/query extensions" if $in_extensions > 0;
```

File: Bugzilla.pm [hunk 5]
```diff
@@ -133,9 +142,20 @@ sub extensions {
       push @$extensions, $package;
     }
   }
+  $recursive = 0;
```
Comment (dylanwh) [inline Bugzilla.pm:145]:
```suggestion
```

File: Bugzilla.pm [hunk 6]
```diff
@@ -124,6 +125,14 @@ sub template_inner {
 }
 
 sub extensions {
+  # Guard against extensions querying the extension list during initialization
+  # (through this method or has_extension).
+  # The extension list is not fully populated at that point,
+  # so the results would not be meaningful.
+  state $recursive = 0;
+  die "Recursive attempt to load/query extensions" if $recursive;
+  $recursive = 1;
```
Comment (dylanwh) [inline Bugzilla.pm:134]:
```suggestion
  local $in_extensions = $in_extensions + 1;
```

Comment (CyberShadow) [reply Bugzilla.pm:134]: Thanks, applied these in https://github.com/bugzilla/harmony/pull/66

**Comment by CyberShadow** [comment]
I have now resubmitted some of these as individual pull requests, and will resubmit the rest once their prerequisites are merged (#65 and #66).

**Pull request closed without merging**